### PR TITLE
Sprite lab limit sprites/texts

### PR DESCRIPTION
**Pull request by fisher-alice**
## Summary
- In Sprite lab, there are two commands, makeNumSprites and makeBurst, which create multiple sprites, but the number is capped behind the scenes. This [PR](https://github.com/code-dot-org/code-dot-org/pull/44075) caps the number of sprites for these two sprite commands. Students can get around these limits by using loops, behaviors, or “while” events which causes the browser to freeze. 
- This PR limits the number of sprites created with early returns in relevant functions once the max is reached. When more than 1000 sprites are created, follow-up work will add a workspace alert so that user will know they reached sprite limit.  
- This PR also updates the printing of the text console statements so that the last 1000 text statements created are printed. Currently, the first 999 created and then then the last statement created is displayed.

## Before Updates
- This video shows the browser freezing when number of sprites is at 100,000:

https://user-images.githubusercontent.com/107423305/185943160-22ddff3d-ca90-4ab4-9aa3-a1ed4970beb1.mp4

- This video shows the browser freezing when makeNumSprites and makeBurst commands are used within a loop despite the current BIG_NUMBER_GUARD:

https://user-images.githubusercontent.com/107423305/185958381-1a210ae5-0ffa-4969-be0b-9e75ec3e3946.mp4

- This video shows that if more than 1000 print text commands are run, the first 999 and then the last print statement is displayed (numbers 1-999, then 1200):

https://user-images.githubusercontent.com/107423305/186692721-f4acae23-5aff-4bb7-b54b-6261413f9bb1.mp4

## After Updates
- This video shows the program is editable even when number of sprites is at 100,000 or greater:

https://user-images.githubusercontent.com/107423305/185953622-1788edd6-1ee9-462f-9af3-0ce2af06c7f5.mp4

- This video shows the program is editable even when makeNumSprites and makeBurst are used within a loop:

https://user-images.githubusercontent.com/107423305/185962483-6b075b63-8d36-48b3-95bd-48a859007fd9.mp4

- This video shows that last 1000 print text statements that were run are displayed (numbers 201-1200):

https://user-images.githubusercontent.com/107423305/186693061-e1408eb8-b036-4049-a2b8-de7d96c19528.mp4

## Links
- [jira ticket - Cap total number of sprites](https://codedotorg.atlassian.net/browse/STAR-2245)
Note that the workspace alert will be included in the follow-up work described below.
- [jira ticket - Prototype: How can we allow users to block the auto-run functionality in Sprite Lab](https://codedotorg.atlassian.net/browse/STAR-2323)

## Testing Story 
- Updated and added unit tests for makeNumSprites
- Added unit tests for makeBurst and printText

## Follow-up work
- The next step is that when a user creates MAX_NUM_SPRITES sprites either during the preview (autorun) or when the program is running, a workspace alert will be displayed with a message. The workspace alert will include the maximum possible number of sprites and a message that additional sprites are not created.

### CODE REVIEW

File: apps/src/p5lab/P5Lab.js [hunk 1]
```diff
@@ -794,7 +794,6 @@ export default class P5Lab {
     clearMarks(DRAW_LOOP_START);
     clearMeasures(DRAW_LOOP_MEASURE);
     this.spriteTotalCount = 0;
-
```
Comment (sanchitmalhotra126) [inline apps/src/p5lab/P5Lab.js]: nit: mind reverting this just so the file isn't included in the PR?

File: apps/src/p5lab/spritelab/CoreLibrary.js [hunk 2]
```diff
@@ -3,6 +3,11 @@ import * as drawUtils from '@cdo/apps/p5lab/drawUtils';
 import commands from './commands/index';
 import {APP_HEIGHT, APP_WIDTH} from '../constants';
 
+// Big numbers in some blocks can cause performance issues. Combined with live-preview,
+// this results in hanging the tab and students unable to edit their blocks. We
+// guard against this by capping numbers where needed.
+const SPRITE_BIG_NUMBER_GUARD = 1000;
```
Comment (sanchitmalhotra126) [inline apps/src/p5lab/spritelab/CoreLibrary.js]: What do you think about storing these constants in a separate file so we can share some of these numbers? I see we reference `SPRITE_BIG_NUMBER_GUARD` in two different places and they have the same value. Maybe some sort of `p5lab/spritelab/constants.js` or `limits.js` or something?

**Review by sanchitmalhotra126 — COMMENTED** [review]
Nice work! Really excited to see this come together :) a few small comments but nothing major.

File: apps/src/p5lab/P5Lab.js (hunk 1 above)
Comment (fisher-alice) [reply apps/src/p5lab/P5Lab.js]: Sure! Will do.

File: apps/src/p5lab/spritelab/CoreLibrary.js (hunk 2 above)
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: Great idea! I will place in the existing spritelab/constants.js file.

**Review by sanchitmalhotra126 — APPROVED** [review]
LGTM!

File: apps/src/p5lab/spritelab/constants.js [hunk 3]
```diff
@@ -97,3 +97,11 @@ export const exampleSprites = {
...
   }
 };
+
+// Big numbers in some blocks can cause performance issues. Combined with live-preview,
+// this results in hanging the tab and students unable to edit their blocks. We
+// guard against this by capping numbers where needed.
+export const BEHAVIOR_BIG_NUMBER_GUARD = 1000;
+export const SPEECH_BUBBLE_BIG_NUMBER_GUARD = 250;
+export const SPRITE_BIG_NUMBER_GUARD = 1000;
+export const TEXT_BIG_NUMBER_GUARD = 250;
```
Comment (breville) [inline apps/src/p5lab/spritelab/constants.js]: I find `BIG_NUMBER` a bit confusing, especially since the new behaviour is really more of a `MAX_` count/instances.

Comment (fisher-alice) [reply apps/src/p5lab/spritelab/constants.js]: Thanks! I agree. Will change constants' names in the following way:  MAX_NUM_SPRITES = 1000, etc.

File: apps/src/p5lab/spritelab/CoreLibrary.js [hunk 4]
```diff
@@ -357,6 +362,9 @@ export default class CoreLibrary {
    */
   addSprite(opts) {
     opts = opts || {};
+    if (this.getNumberOfSprites() >= MAX_NUM_SPRITES) {
```
Comment (mikeharv) [inline apps/src/p5lab/spritelab/CoreLibrary.js:374]: We discussed moving this comparison to a helper function and using it here as well as in spriteCommands.

File: apps/src/p5lab/spritelab/commands/spriteCommands.js [hunk 5]
```diff
@@ -70,7 +66,11 @@ export const commands = {
   },
 
   makeNumSprites(num, animation) {
-    num = Math.min(num, BIG_NUMBER_GUARD);
+    var numSprites = this.getNumberOfSprites();
+    if (numSprites >= MAX_NUM_SPRITES) {
```
Comment (mikeharv) [inline apps/src/p5lab/spritelab/commands/spriteCommands.js]: This is another place where we could use a helper function.

File: apps/src/p5lab/spritelab/commands/spriteCommands.js [hunk 6]
```diff
@@ -80,13 +80,17 @@ export const commands = {
   },
 
   makeBurst(num, animation, effectName) {
+    var numSprites = this.getNumberOfSprites();
+    if (numSprites >= MAX_NUM_SPRITES) {
```
Comment (mikeharv) [inline apps/src/p5lab/spritelab/commands/spriteCommands.js]: Ditto here. We can use the same helper function.

File: apps/src/p5lab/spritelab/commands/spriteCommands.js [hunk 7]
```diff
@@ -70,7 +66,11 @@ export const commands = {
   },
 
   makeNumSprites(num, animation) {
-    num = Math.min(num, BIG_NUMBER_GUARD);
+    var numSprites = this.getNumberOfSprites();
+    if (numSprites >= MAX_NUM_SPRITES) {
+      return;
+    }
+    num = Math.min(num, MAX_NUM_SPRITES - numSprites);
```
Comment (mikeharv) [inline apps/src/p5lab/spritelab/commands/spriteCommands.js]: This is great!

File: apps/src/p5lab/spritelab/CoreLibrary.js [hunk 8]
```diff
@@ -343,6 +344,10 @@ export default class CoreLibrary {
     return spriteIds;
   }
 
+  getNumberOfSprites() {
```
Comment (mikeharv) [inline apps/src/p5lab/spritelab/CoreLibrary.js:347]: This is a very useful helper!

File: apps/src/p5lab/spritelab/commands/spriteCommands.js [hunk 9]
```diff
@@ -80,13 +80,17 @@ export const commands = {
...
+      return;
+    }
     const behaviorFuncs = {
       burst: behaviorCommands.burstFunc,
       pop: behaviorCommands.popFunc,
       rain: behaviorCommands.rainFunc,
       spiral: behaviorCommands.spiralFunc
     };
-    num = Math.min(num, BIG_NUMBER_GUARD);
+    num = Math.min(num, MAX_NUM_SPRITES);
```
Comment (mikeharv) [inline apps/src/p5lab/spritelab/commands/spriteCommands.js]: Could we subtract `numSprites` here like in makeNumSprites?

```
num = Math.min(num, MAX_NUM_SPRITES - numSprites);
```

File: apps/src/p5lab/spritelab/commands/spriteCommands.js [hunk 10]
```diff
@@ -70,7 +66,11 @@ export const commands = {
   },
 
   makeNumSprites(num, animation) {
-    num = Math.min(num, BIG_NUMBER_GUARD);
+    var numSprites = this.getNumberOfSprites();
```
Comment (mikeharv) [inline apps/src/p5lab/spritelab/commands/spriteCommands.js]: Nit: replace `var` with `const`

File: apps/src/p5lab/spritelab/commands/spriteCommands.js [hunk 11]
```diff
@@ -80,13 +80,17 @@ export const commands = {
   },
 
   makeBurst(num, animation, effectName) {
+    var numSprites = this.getNumberOfSprites();
```
Comment (mikeharv) [inline apps/src/p5lab/spritelab/commands/spriteCommands.js]: Nit: replace `var` with `const`

**Review by mikeharv — CHANGES_REQUESTED** [review]
Great work here, Alice!
As we discussed, the next step is to create a helper function to be used to determine if we should return early as well as to "cap" the value of `num` so we don't have to go through any of the loops more times than needed. Overall, this feels like the right solution, so you should also be good to start removing any commented out code.

File: apps/src/p5lab/spritelab/CoreLibrary.js (hunk 4 above)
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/CoreLibrary.js:374]: Since this is only adding a single sprite, I think that this comparison suffices. But there are two other comparisons where I've updated the code with the helper function - thanks!

File: apps/src/p5lab/spritelab/commands/spriteCommands.js (hunk 5 above)
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/commands/spriteCommands.js]: Thanks - made the update!

File: apps/src/p5lab/spritelab/commands/spriteCommands.js (hunk 9 above)
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/commands/spriteCommands.js]: Sure! The helper function definitely helps prevent this kind of typo!

File: apps/src/p5lab/spritelab/commands/spriteCommands.js (hunk 10 above)
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/commands/spriteCommands.js]: Thanks - made the update!

File: apps/src/p5lab/spritelab/CoreLibrary.js (hunk 4 above)
Comment (mikeharv) [reply apps/src/p5lab/spritelab/CoreLibrary.js:374]: Given the purpose of the new helper function, I agree.

**Review by mikeharv — APPROVED** [review]
Awesome work! Once we have completed the additional work to add workspace alerts, this looks good to merge!

File: apps/src/p5lab/spritelab/commands/spriteCommands.js [hunk 12]
```diff
@@ -70,8 +65,12 @@ export const commands = {
   },
 
   makeNumSprites(num, animation) {
-    num = Math.min(num, BIG_NUMBER_GUARD);
-    for (let i = 0; i < num; i++) {
+    // getCappedNumSprites caps num based on number of sprites already created and MAX_NUM_SPRITES
+    const cappedNumSprites = this.getCappedNumSprites(num);
```
Comment (sanchitmalhotra126) [inline apps/src/p5lab/spritelab/commands/spriteCommands.js]: minor nit but maybe this should be called `maxAllowedNewSprites/getMaxAllowedNewSprites`, or something along those lines?

**Review by sanchitmalhotra126 — APPROVED** [review]
One minor comment, but otherwise LGTM!

File: apps/src/p5lab/spritelab/commands/spriteCommands.js [hunk 13]
```diff
@@ -70,8 +65,12 @@ export const commands = {
   },
 
   makeNumSprites(num, animation) {
```
Comment (jmkulwik) [inline apps/src/p5lab/spritelab/commands/spriteCommands.js]: NIT: Seeing "num" in several places starts to get a little hard to follow. It indicates that it's storing a number, but it's not quite clear what that number is used for. I'd recommend renaming it to `numSprites`, `spriteCount`, `numSpritesRequested` or something similar.

**Comment by jmkulwik** [comment]
Great documentation! It's so awesome to get this issue fixed for our students!!

**Review by jmkulwik — APPROVED** [review]

File: apps/test/unit/p5lab/spritelab/spriteCommandsTest.js [hunk 14]
```diff
@@ -140,11 +140,11 @@ describe('Sprite Commands', () => {
       ).to.equal(10);
     });
 
-    it('caps at 500 sprites', () => {
+    it('caps at 1000 sprites', () => {
```
Comment (jmkulwik) [inline apps/test/unit/p5lab/spritelab/spriteCommandsTest.js]: There's some great functionality added here. To ensure it keeps working in the long-term, add a few more tests to cover added scenarios. For example:
* display only the last 1000 lines of text
* calling makeNumSprites 3 times with 500 sprites should still cap at 1000 sprites

**Review by jmkulwik — COMMENTED** [review]
Left a comment for a few more tests. Once those are in, LGTM!

File: apps/src/p5lab/spritelab/commands/spriteCommands.js (hunk 12 above)
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/commands/spriteCommands.js]: Sounds good - will update.

File: apps/src/p5lab/spritelab/commands/spriteCommands.js (hunk 13 above)
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/commands/spriteCommands.js]: I agree! Will update.

File: apps/src/p5lab/spritelab/commands/spriteCommands.js (hunk 11 above)
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/commands/spriteCommands.js]: Thanks - made the update!

File: apps/test/unit/p5lab/spritelab/spriteCommandsTest.js (hunk 14 above)
Comment (fisher-alice) [reply apps/test/unit/p5lab/spritelab/spriteCommandsTest.js]: Thanks - great suggestion! Added additional unit test for makeNumSprites and added tests for makeBurst and printText.
